File: apps/android/app/src/main/kotlin/com/epicstartup/tenant/core/net/TenantApiClient.kt
package com.epicstartup.tenant.core.net

import com.epicstartup.tenant.core.config.TenantConfiguration
import com.epicstartup.tenant.core.model.AuthTokens
import com.epicstartup.tenant.core.model.CustomerProfile
import com.epicstartup.tenant.core.model.PublicOrganization
import com.epicstartup.tenant.core.model.VerifyResult
import com.epicstartup.tenant.core.support.objectOrNull
import com.epicstartup.tenant.core.support.parseJsonObject
import java.net.URLEncoder
import org.json.JSONObject

/**
 * Typed client for the two public APIs the tenant app depends on:
 *
 * - **US control-plane App** (`/resources/sites`) for published org branding.
 *   No customer PII.
 * - **Regional tenant-api** (`/auth/…`) for phone-OTP customer sessions and the
 *   customer profile. This is the only place customer PII lives, and the app
 *   talks to it directly — nothing is proxied through the control plane.
 *
 * Every call blocks; the UI layer runs them on a background executor.
 */
class TenantApiClient(
	val configuration: TenantConfiguration,
	private val transport: HttpTransport = HttpUrlConnectionTransport(),
) {

	// MARK: - Published branding (App)

	fun fetchOrganization(locale: String? = null, acceptLanguage: String? = null): PublicOrganization {
		val query = configuration.siteAddress.queryItems.toMutableList()
		if (query.isEmpty()) throw ApiException.notBoundToTenant()
		if (!locale.isNullOrEmpty()) query += "lng" to locale

		val headers = mutableMapOf("Accept" to "application/json")
		// The App negotiates the response locale from Accept-Language, the same
		// way URLSession does for the iOS app.
		if (!acceptLanguage.isNullOrBlank()) headers["Accept-Language"] = acceptLanguage

		val request = buildRequest(
			baseUrl = configuration.appBaseUrl,
			path = "/resources/sites",
			query = query,
		).copy(headers = headers)

		val body = perform(request)
		val json = parseJsonObject(body)
			?: throw ApiException(200, "Unexpected response from /resources/sites")
		return PublicOrganization.fromJson(json)
			?: throw ApiException(200, "Unexpected response from /resources/sites")
	}

	// MARK: - Customer auth (regional tenant-api)

	fun sendCode(phone: String) {
		val payload = JSONObject().apply {
			put("phone", phone)
			applyBinding(this)
		}
		val request = buildRequest(
			baseUrl = configuration.tenantApiBaseUrl,
			path = "/auth/send-code",
			query = emptyList(),
		).copy(
			method = "POST",
			headers = jsonHeaders(withBindingOrigin = true),
			body = payload.toString().toByteArray(Charsets.UTF_8),
		)
		perform(request)
	}

	fun verify(phone: String, code: String): VerifyResult {
		val payload = JSONObject().apply {
			put("phone", phone)
			put("code", code)
			applyBinding(this)
		}
		val request = buildRequest(
			baseUrl = configuration.tenantApiBaseUrl,
			path = "/auth/verify",
			query = emptyList(),
		).copy(
			method = "POST",
			headers = jsonHeaders(withBindingOrigin = true),
			body = payload.toString().toByteArray(Charsets.UTF_8),
		)

		val result = send(request, "/auth/verify") { VerifyResult.fromJson(it) }
		if (result.accessToken.isNullOrEmpty()) {
			throw ApiException(400, result.error ?: "Verification failed")
		}
		return result
	}

	fun refresh(refreshToken: String, orgId: String): AuthTokens {
		val payload = JSONObject().apply {
			put("refreshToken", refreshToken)
			put("orgId", orgId)
		}
		val request = buildRequest(
			baseUrl = configuration.tenantApiBaseUrl,
			path = "/auth/refresh",
			query = emptyList(),
		).copy(
			method = "POST",
			headers = jsonHeaders(),
			body = payload.toString().toByteArray(Charsets.UTF_8),
		)

		val result = send(request, "/auth/refresh") { VerifyResult.fromJson(it) }
		val accessToken = result.accessToken
		if (accessToken.isNullOrEmpty()) {
			throw ApiException(401, "Session expired. Please sign in again.")
		}
		return AuthTokens(accessToken = accessToken, refreshToken = result.refreshToken)
	}

	fun logout(refreshToken: String?, orgId: String?) {
		if (refreshToken == null || orgId == null) return
		val payload = JSONObject().apply {
			put("refreshToken", refreshToken)
			put("orgId", orgId)
		}
		val request = buildRequest(
			baseUrl = configuration.tenantApiBaseUrl,
			path = "/auth/logout",
			query = emptyList(),
		).copy(
			method = "POST",
			headers = jsonHeaders(),
			body = payload.toString().toByteArray(Charsets.UTF_8),
		)
		runCatching { perform(request) }
	}

	fun fetchProfile(accessToken: String): CustomerProfile {
		val request = buildRequest(
			baseUrl = configuration.tenantApiBaseUrl,
			path = "/auth/me",
			query = emptyList(),
		).copy(
			headers = mapOf(
				"Accept" to "application/json",
				"Authorization" to "Bearer $accessToken",
			),
		)

		val envelope = parseJsonObject(perform(request))
			?: throw ApiException(200, "Unexpected response from /auth/me")
		val customer = envelope.objectOrNull("customer")
			?: throw ApiException(404, "Customer not found")
		return CustomerProfile.fromJson(customer)
	}

	fun updateProfile(name: String, email: String?, accessToken: String): AuthTokens {
		val payload = JSONObject().apply {
			put("name", name)
			// `null` means "leave the stored email alone"; an empty string clears it.
			if (email != null) put("email", email)
		}
		val request = buildRequest(
			baseUrl = configuration.tenantApiBaseUrl,
			path = "/auth/profile",
			query = emptyList(),
		).copy(
			method = "POST",
			headers = mapOf(
				"Accept" to "application/json",
				"Content-Type" to "application/json",
				"Authorization" to "Bearer $accessToken",
			),
			body = payload.toString().toByteArray(Charsets.UTF_8),
		)

		val result = send(request, "/auth/profile") { VerifyResult.fromJson(it) }
		val nextAccessToken = result.accessToken
		if (nextAccessToken.isNullOrEmpty()) {
			throw ApiException(400, result.error ?: "Could not save your profile")
		}
		return AuthTokens(accessToken = nextAccessToken, refreshToken = result.refreshToken)
	}

	// MARK: - Request plumbing

	/**
	 * tenant-api binds `send-code`/`verify` to the tenant the caller claims to
	 * be: slug or host in the body, plus the tenant's public site origin for
	 * https (production) origins — the same origin↔org rule Sites satisfies via
	 * its Host header. Local dev omits the header and relies on the body binding
	 * that tenant-api allows outside production.
	 */
	private fun applyBinding(payload: JSONObject) {
		for ((name, value) in configuration.siteAddress.queryItems) {
			payload.put(name, value)
		}
	}

	private fun jsonHeaders(withBindingOrigin: Boolean = false): Map<String, String> {
		val headers = mutableMapOf(
			"Accept" to "application/json",
			"Content-Type" to "application/json",
		)
		if (withBindingOrigin) {
			configuration.siteAddress.authOriginHeader?.let { headers["Origin"] = it }
		}
		return headers
	}

	private fun buildRequest(
		baseUrl: String,
		path: String,
		query: List<Pair<String, String>>,
	): ApiRequest {
		val url = buildString {
			append(baseUrl.trimEnd('/'))
			append(path)
			if (query.isNotEmpty()) {
				append('?')
				append(
					query.joinToString("&") { (name, value) ->
						"${encode(name)}=${encode(value)}"
					},
				)
			}
		}
		return ApiRequest(url = url, method = "GET", timeoutMs = configuration.requestTimeoutMs)
	}

	private fun encode(value: String): String = URLEncoder.encode(value, "UTF-8")

	private fun perform(request: ApiRequest): ByteArray {
		val response = try {
			transport.send(request)
		} catch (error: ApiException) {
			throw error
		} catch (error: Exception) {
			throw ApiException(0, error.message ?: "Network error")
		}
		if (response.statusCode !in 200..299) {
			throw ApiException.from(response.statusCode, response.body)
		}
		return response.body
	}

	private fun <T> send(request: ApiRequest, path: String, decode: (JSONObject) -> T): T {
		val body = perform(request)
		val json = parseJsonObject(body)
			?: throw ApiException(200, "Unexpected response from $path")
		return decode(json)
	}
}

File: apps/ios/Sources/TenantKit/Networking/TenantAPIClient.swift
import Foundation

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// Typed client for the two public APIs the tenant app depends on:
///
/// - **US control-plane App** (`/resources/sites`) for published org branding.
///   No customer PII.
/// - **Regional tenant-api** (`/auth/*`) for phone-OTP customer sessions and the
///   customer profile. This is the only place customer PII lives, and the app
///   talks to it directly — nothing is proxied through the control plane.
public struct TenantAPIClient: Sendable {
	public let configuration: TenantConfiguration
	private let transport: HTTPTransport

	public init(
		configuration: TenantConfiguration,
		transport: HTTPTransport = URLSessionTransport()
	) {
		self.configuration = configuration
		self.transport = transport
	}

	// MARK: - Published branding (App)

	public func fetchOrganization(locale: String? = nil) async throws -> PublicOrganization {
		var items = configuration.siteAddress.queryItems
		guard !items.isEmpty else { throw APIError.notBoundToTenant }
		if let locale, !locale.isEmpty {
			items.append(URLQueryItem(name: "lng", value: locale))
		}

		var request = try makeRequest(
			base: configuration.appBaseURL,
			path: "/resources/sites",
			queryItems: items
		)
		request.httpMethod = "GET"
		request.setValue("application/json", forHTTPHeaderField: "Accept")
		return try await send(request, as: PublicOrganization.self)
	}

	// MARK: - Customer auth (regional tenant-api)

	public func sendCode(phone: String) async throws {
		var body: [String: String] = ["phone": phone]
		applyBinding(to: &body)
		var request = try makeRequest(base: configuration.tenantAPIBaseURL, path: "/auth/send-code", queryItems: [])
		request.httpMethod = "POST"
		request.setValue("application/json", forHTTPHeaderField: "Accept")
		request.setValue("application/json", forHTTPHeaderField: "Content-Type")
		applyBindingHeaders(to: &request)
		request.httpBody = try JSONSerialization.data(withJSONObject: body)
		_ = try await perform(request)
	}

	public func verify(phone: String, code: String) async throws -> VerifyResult {
		var body: [String: String] = ["phone": phone, "code": code]
		applyBinding(to: &body)
		var request = try makeRequest(base: configuration.tenantAPIBaseURL, path: "/auth/verify", queryItems: [])
		request.httpMethod = "POST"
		request.setValue("application/json", forHTTPHeaderField: "Accept")
		request.setValue("application/json", forHTTPHeaderField: "Content-Type")
		applyBindingHeaders(to: &request)
		request.httpBody = try JSONSerialization.data(withJSONObject: body)

		let result = try await send(request, as: VerifyResult.self)
		guard let accessToken = result.accessToken, !accessToken.isEmpty else {
			throw APIError(statusCode: 400, message: result.error ?? "Verification failed")
		}
		return result
	}

	public func refresh(refreshToken: String, orgId: String) async throws -> AuthTokens {
		var request = try makeRequest(base: configuration.tenantAPIBaseURL, path: "/auth/refresh", queryItems: [])
		request.httpMethod = "POST"
		request.setValue("application/json", forHTTPHeaderField: "Accept")
		request.setValue("application/json", forHTTPHeaderField: "Content-Type")
		request.httpBody = try JSONSerialization.data(withJSONObject: [
			"refreshToken": refreshToken,
			"orgId": orgId,
		])

		let payload = try await send(request, as: VerifyResult.self)
		guard let accessToken = payload.accessToken, !accessToken.isEmpty else {
			throw APIError(statusCode: 401, message: "Session expired. Please sign in again.")
		}
		return AuthTokens(accessToken: accessToken, refreshToken: payload.refreshToken)
	}

	public func logout(refreshToken: String?, orgId: String?) async {
		guard let refreshToken, let orgId else { return }
		guard
			var request = try? makeRequest(base: configuration.tenantAPIBaseURL, path: "/auth/logout", queryItems: [])
		else { return }
		request.httpMethod = "POST"
		request.setValue("application/json", forHTTPHeaderField: "Content-Type")
		request.httpBody = try? JSONSerialization.data(withJSONObject: [
			"refreshToken": refreshToken,
			"orgId": orgId,
		])
		_ = try? await perform(request)
	}

	public func fetchProfile(accessToken: String) async throws -> CustomerProfile {
		var request = try makeRequest(base: configuration.tenantAPIBaseURL, path: "/auth/me", queryItems: [])
		request.httpMethod = "GET"
		request.setValue("application/json", forHTTPHeaderField: "Accept")
		request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
		let envelope = try await send(request, as: CustomerProfileEnvelope.self)
		guard let customer = envelope.customer else {
			throw APIError(statusCode: 404, message: "Customer not found")
		}
		return customer
	}

	public func updateProfile(
		name: String,
		email: String?,
		accessToken: String
	) async throws -> AuthTokens {
		var request = try makeRequest(base: configuration.tenantAPIBaseURL, path: "/auth/profile", queryItems: [])
		request.httpMethod = "POST"
		request.setValue("application/json", forHTTPHeaderField: "Accept")
		request.setValue("application/json", forHTTPHeaderField: "Content-Type")
		request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
		// `nil` means "leave the stored email alone"; an empty string clears it.
		var body: [String: String] = ["name": name]
		if let email {
			body["email"] = email
		}
		request.httpBody = try JSONSerialization.data(withJSONObject: body)

		let payload = try await send(request, as: VerifyResult.self)
		guard let nextAccessToken = payload.accessToken, !nextAccessToken.isEmpty else {
			throw APIError(statusCode: 400, message: payload.error ?? "Could not save your profile")
		}
		return AuthTokens(accessToken: nextAccessToken, refreshToken: payload.refreshToken)
	}

	// MARK: - Request plumbing

	/// tenant-api binds `send-code`/`verify` to the tenant the caller claims to
	/// be: slug or host in the body, plus the tenant's public site origin for
	/// https (production) origins — the same origin↔org rule Sites satisfies via
	/// its Host header. Local dev omits the header and relies on the body binding
	/// that tenant-api allows outside production.
	private func applyBinding(to body: inout [String: String]) {
		for item in configuration.siteAddress.queryItems {
			body[item.name] = item.value
		}
	}

	private func applyBindingHeaders(to request: inout URLRequest) {
		if let origin = configuration.siteAddress.authOriginHeader {
			request.setValue(origin, forHTTPHeaderField: "Origin")
		}
	}

	private func makeRequest(
		base: URL,
		path: String,
		queryItems: [URLQueryItem]
	) throws -> URLRequest {
		let baseURL = base.appendingPathComponent(path)
		guard var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false) else {
			throw APIError(statusCode: 0, message: "Invalid URL for \(path)")
		}
		if !queryItems.isEmpty {
			components.queryItems = queryItems
		}
		guard let url = components.url else {
			throw APIError(statusCode: 0, message: "Invalid URL for \(path)")
		}
		var request = URLRequest(url: url)
		request.timeoutInterval = configuration.requestTimeout
		return request
	}

	private func perform(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
		do {
			let (data, response) = try await transport.send(request)
			guard (200..<300).contains(response.statusCode) else {
				throw APIError.from(statusCode: response.statusCode, data: data)
			}
			return (data, response)
		} catch let error as APIError {
			throw error
		} catch {
			throw APIError(statusCode: 0, message: error.localizedDescription)
		}
	}

	private func send<T: Decodable>(_ request: URLRequest, as type: T.Type) async throws -> T {
		let (data, response) = try await perform(request)
		do {
			return try JSONDecoder().decode(T.self, from: data)
		} catch {
			throw APIError(
				statusCode: response.statusCode,
				message: "Unexpected response from \(response.url?.path ?? "the server")"
			)
		}
	}
}
